lib/event_store_client/adapters/grpc/cluster/gossip_discover.rb

```diff
@@ -0,0 +1,131 @@
+# frozen_string_literal: true
+
+# rubocop:disable Metrics/AbcSize, Metrics/MethodLength
+
+require 'event_store_client/adapters/grpc/generated/shared_pb'
+require 'event_store_client/adapters/grpc/generated/gossip_pb'
+require 'event_store_client/adapters/grpc/generated/gossip_services_pb'
+
+module EventStoreClient
+  module GRPC
+    module Cluster
+      class GossipDiscover
+        include Configuration
+
+        DiscoverError = Class.new(StandardError)
+
+        # Order is important - it plays role of states priority as well
+        READ_ONLY_STATES = %i[ReadOnlyReplica PreReadOnlyReplica ReadOnlyLeaderless].freeze
+        ALLOWED_NODE_STATES =
+          (%i[Leader Follower] + READ_ONLY_STATES).freeze
+
+        # @param nodes [Array<EventStoreClient::Connection::Url::Node>]
+        # @param failed_member [EventStoreClient::GRPC::Cluster::Member, nil]
+        # @return [EventStoreClient::GRPC::Cluster::Member]
+        # @raise [EventStoreClient::GRPC::Cluster::GossipDiscover::DiscoverError]
+        def call(nodes, failed_member:)
+          nodes = sort_nodes(nodes, failed_member)
+
+          attempts = config.eventstore_url.max_discover_attempts
+          attempts.times do
+            nodes.each do |node|
+              suitable_member = suitable_member_of_node(node)
+              next unless suitable_member
+
+              return suitable_member
+            end
+
+            sleep(config.eventstore_url.discover_interval / 1000.0)
+          end
+
+          raise DiscoverError, "Failed to discover suitable host after #{attempts} attempts."
+        end
+
+        private
+
+        # @return [Array<Symbol>]
+        def ordered_states
+          @ordered_states ||=
+            # Move preferred state to the first place
+            case config.eventstore_url.node_preference
+            when :Leader, :Follower
+              [config.eventstore_url.node_preference] +
+              (ALLOWED_NODE_STATES - [config.eventstore_url.node_preference])
```

Comment:
These array operations you can hide behind some private method.

Reply:
I don't want to hide them :) It is already simple and it doesn't worth its own method